base/cartesian.jl

```diff
@@ -257,6 +257,57 @@ macro nif(N, condition, operation...)
     ex
 end
 
+"""
+    nif(condition, expression, ::Val{N}) where {N}
+    nif(condition, expression, else_expression, ::Val{N}) where {N}
+
+Generate a sequence of `if ... elseif ... else ... end` statements.
+
+# Arguments
+- `condition`: A function that takes an integer between `1` and `N-1` and
+    returns a boolean condition.
+- `expression`: A function that takes an integer between `1` and `N` (or,
+    only up to `N-1`, if `else_expression` is provided) and is called if
+    the condition is true.
+- `else_expression`: (optional) A function that takes `N` as input
+    returns an expression to be evaluated if all conditions are false.
+- `N`: The number of conditions to check, passed as a `Val{N}` type.
+
+This function is similar to the `@nif` macro but can be used in cases
+where `N` is not known at parse time.
+
+# Examples
+
+For example, here we find the first index of a positive element in a
+fixed-size tuple using `nif`:
+
+```jldoctest
+julia> x = (0, -1, 1, 0)
+(0, -1, 1, 0)
+
+julia> Base.Cartesian.nif(d -> x[d] > 0, d -> d, Val(4))
+3
+```
+"""
+@inline function nif(condition, expression, ::Val{N}) where {N}
+    nif(condition, expression, expression, Val(N))
+end
+@inline function nif(condition::F, expression::G, else_expression::H, ::Val{N}) where {F,G,H,N}
+    n = N::Int  # Can improve inference; see #54544
+    (n >= 0) || throw(ArgumentError(LazyString("if statement length should be ≥ 0, got ", n)))
+    if @generated
+        return :(@nif $n d -> condition(d) d -> expression(d) d -> else_expression(d))
```

Comment:
```suggestion
        return :(@nif $N d -> condition(d) d -> expression(d) d -> else_expression(d))
```
The generated branch is only taken if all argument types are fully known, so no need to use the type-asserted `n` here, and for some reason, using `n` seems to upset the compiler.

Reply:
Ah, `n` doesn't work because of #55101, I guess.

Reply:
Thanks. Yeah, that is strange.